RealmSwift/Realm.swift

```diff
@@ -1205,42 +1230,208 @@ extension Realm {
     public init(configuration: Realm.Configuration = .defaultConfiguration,
                 downloadBeforeOpen: OpenBehavior = .never) async throws {
         let scheduler = RLMScheduler.dispatchQueue(.main)
-        let rlmConfiguration = configuration.rlmConfiguration
-
-        // If we already have a cached Realm for this actor, just reuse it
-        // If this Realm is open but with a different scheduler, open it synchronously.
-        // An async open would just dispatch to the background and then back to
-        // perform the final synchronous open.
-        var realm = RLMGetCachedRealm(rlmConfiguration, scheduler)
-        if realm == nil, let cachedRealm = RLMGetAnyCachedRealm(rlmConfiguration) {
-            realm = try withExtendedLifetime(cachedRealm) {
-                try RLMRealm(configuration: rlmConfiguration, confinedTo: scheduler)
-            }
+        let rlmRealm = try await openRealm(configuration: configuration, scheduler: scheduler,
+                                           actor: MainActor.shared, downloadBeforeOpen: downloadBeforeOpen)
+        self = Realm(rlmRealm.wrappedValue)
+    }
+
+#if swift(>=5.8)
+    /**
+     Asynchronously obtains a `Realm` instance isolated to the given Actor.
+
+     Opening a Realm with an actor isolates the Realm to that actor. Rather
+     than being confined to the specific thread which the Realm was opened on,
+     the Realm can instead only be used from within that actor or functions
+     isolated to that actor. Isolating a Realm to an actor also enables using
+     ``asyncWrite`` and ``asyncRefresh``.
+
+     All initialization work to prepare the Realm for work, such as creating,
+     migrating, or compacting the file on disk, and waiting for synchronized
+     Realms to download the latest data from the server is done on a background
+     thread and does not block the calling executor.
+
+     When using actor-isolated Realms, enabling struct concurrency checking
+     (`SWIFT_STRICT_CONCURRENCY=complete` in Xcode) and runtime data race
+     detection (by passing `-Xfrontend -enable-actor-data-race-checks` to the
+     compiler) is strongly recommended.
+
+     - parameter configuration: A configuration object to use when opening the Realm.
+     - parameter actor: The actor to confine this Realm to. The actor can be
+     either a local actor or a global actor. The calling function does not need
+     to be isolated to the actor passed in, but if it is not it will not be
+     able to use the returned Realm.
+     - parameter downloadBeforeOpen: When opening the Realm should first download
+     all data from the server.
+     - throws: An `NSError` if the Realm could not be initialized.
+               `CancellationError` if the task is cancelled.
+     - returns: An open Realm.
+     */
+    public init<A: Actor>(configuration: Realm.Configuration = .defaultConfiguration,
```

Comment:
Adding a note her as per our discussion: it would be nice to have a synchronous init specifically for the case of synchronous actor initialization.

Reply:
I think this needs SE-0392 (custom actor executors) to be reasonable. With the current limitations it'd have to be entirely unchecked and would be a super error-prone API.